src/Lecture2.hs

```diff
@@ -264,8 +356,25 @@ data EvalError
 It returns either a successful evaluation result or an error.
 -}
 eval :: Variables -> Expr -> Either EvalError Int
-eval = error "TODO"
-
+eval _ (Lit i) = Right i
+eval vars (Var s) =
+  let
+    val = lookup s vars
+  in
+    if isJust val
+    then Right (fromJust val)
+    else Left (VariableNotFound s)
```

Comment:
You're using the partial function `fromJust` here and you compare your value with `Nothing` first using `isJust`. It's correct, however, such an approach is considered an antipattern. You can use `case-of` instead to write the correct-by-construction function:

```suggestion
    case val of
        Nothing -> Left (VariableNotFound x)
        Just n  -> Right n
```